### PR TITLE
issue: 825052 enhance sw latency budget measurements

### PR DESCRIPTION
**Pull request by guy101**

### CODE REVIEW

**Comment by guy101** [comment]
this PR depends on  pr #50  - issue: 825052 Enhance SW latency budget measurements

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/430/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/446/ for details (Mellanox internal link).

File: src/vma/util/flow_instrumentation.h [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+/*
+ * VMA instrumental - measure the times that certain actions takes.
+ * Currently support: RX,TX,IOMUX
+ * Enable: use --enable-time_measure in ./configure
```
Comment (rosenbaumalex) [inline src/vma/util/flow_instrumentation.h]: suggesting to use explaination as: `Enable: ./configure --enable-time_measure`

File: src/vma/util/flow_instrumentation.h [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+ * Parameters:
+ * 	VMA_TIME_MEASURE_DUMP_FILE - Name of the results file. Default: /tmp/VMA_inst.dump
+ * 	VMA_TIME_MEASURE_NUM_SAMPLES - Number of samples for a dump file. Default: 10000
+ * Limitations:
+ * 	- No support for multi-threading
+ * 	- Support only one socket
+ */
+
+#ifndef V_FLOWINSTRUMENTATION_H
```
Comment (rosenbaumalex) [inline src/vma/util/flow_instrumentation.h]: remove the `V_` in define (old name tag from Voltaire era)

File: src/vma/util/flow_instrumentation.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+#define V_FLOWINSTRUMENTATION_H
+
+// RECIEVE FLOW
+ #define FLOW_RECV_TO_CGMGR_RING
```
Comment (rosenbaumalex) [inline src/vma/util/flow_instrumentation.h]: typo in FLOW_RECV_TO_CGMGR_RING? `_CGMGR_` -> `_CQ_MGR_`

File: src/vma/util/instrumentation.h [hunk 4]
```diff
@@ -46,6 +46,8 @@
 #ifndef V_INSTRUMENTATION_H
 #define V_INSTRUMENTATION_H
 
+
+#include "flow_instrumentation.h"
```
Comment (rosenbaumalex) [inline src/vma/util/instrumentation.h]: please move project/module include below the system inclides

File: src/vma/util/flow_instrumentation.h (hunk 1 above)
Comment (igor-ivanov) [inline src/vma/util/flow_instrumentation.h]: @guy101 consider another short name as `--enable-profiling`

File: src/vma/util/flow_instrumentation.h (hunk 2 above)
Comment (igor-ivanov) [inline src/vma/util/flow_instrumentation.h]: @guy101 can we use single file for such things as util/instrumentation.h?

File: src/vma/dev/cq_mgr.cpp [hunk 5]
```diff
@@ -730,9 +730,16 @@ int cq_mgr::poll_and_process_helper_rx(uint64_t* p_cq_poll_sn, void* pv_fd_ready
 	if (m_p_next_rx_desc_poll) {
 		prefetch_range((uint8_t*)m_p_next_rx_desc_poll->p_buffer, m_n_sysvar_rx_prefetch_bytes_before_poll);
 	}
-
+	INSTRUMENT_END_RECV_TO_CGMGR_RING
+	INSTRUMENT_START_CQ_MGR_POLL_WRAPPER
+//double tm_start = ibprof_timestamp();
```
Comment (igor-ivanov) [inline src/vma/dev/cq_mgr.cpp]: should be removed

File: src/vma/dev/cq_mgr.cpp [hunk 6]
```diff
@@ -730,9 +730,16 @@ int cq_mgr::poll_and_process_helper_rx(uint64_t* p_cq_poll_sn, void* pv_fd_ready
...
 	ret = poll(wce, m_n_sysvar_cq_poll_batch_max, p_cq_poll_sn);
+//ibprof_update(IBPROF_MODULE_USER, 55, ibprof_timestamp() - (tm_start));
```
Comment (igor-ivanov) [inline src/vma/dev/cq_mgr.cpp]: should be removed

File: src/vma/sock/sock-redirect.cpp [hunk 7]
```diff
@@ -1480,8 +1514,12 @@ int pselect(int __nfds,
 	} else {
 		srdr_logfunc_entry("nfds=%d, timeout=(infinite)", __nfds);
 	}
-
-	return select_helper(__nfds, __readfds, __writefds, __errorfds, __timeout ? &select_time : NULL, __sigmask);
+	INSTRUMENT_START_RECV_TO_CGMGR_RING
+	int res = select_helper(__nfds, __readfds, __writefds, __errorfds, __timeout ? &select_time : NULL, __sigmask);
+	INSTRUMENT_END_CQ_MGR_MISS_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_TCP_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_UDP_TO_RECV
```
Comment (igor-ivanov) [inline src/vma/sock/sock-redirect.cpp]: please correct alignment

File: src/vma/sock/sock-redirect.cpp [hunk 8]
```diff
@@ -1524,13 +1563,19 @@ int poll(struct pollfd *__fds, nfds_t __nfds, int __timeout)
...
 	else
 		srdr_logfunc_entry("nfds=%d, timeout=(%d milli-sec)", __nfds, __timeout);
-
-	return poll_helper(__fds, __nfds, __timeout);
+	INSTRUMENT_START_RECV_TO_CGMGR_RING
+	int res = poll_helper(__fds, __nfds, __timeout);
+	INSTRUMENT_END_CQ_MGR_MISS_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_TCP_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_UDP_TO_RECV
+	return res;
```
Comment (igor-ivanov) [inline src/vma/sock/sock-redirect.cpp]: please correct alignment

File: src/vma/sock/sock-redirect.cpp [hunk 9]
```diff
@@ -1546,7 +1591,12 @@ int ppoll(struct pollfd *__fds, nfds_t __nfds, const struct timespec *__timeout,
 	else
 		srdr_logfunc_entry("nfds=%d, timeout=(%d milli-sec)", __nfds, timeout);
 
-	return poll_helper(__fds, __nfds, timeout, __sigmask);
+	int res = poll_helper(__fds, __nfds, timeout, __sigmask);
+	INSTRUMENT_END_CQ_MGR_MISS_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_TCP_TO_RECV
+        INSTRUMENT_END_RX_INPUT_CB_UDP_TO_RECV
+
```
Comment (igor-ivanov) [inline src/vma/sock/sock-redirect.cpp]: please correct alignment

File: src/vma/util/flow_instrumentation.h [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+#else
+        #define INSTRUMENT_START_RING_SIMPLE_SEND_RING_BUFFER_TO_SEND
+        #define INSTRUMENT_END_RING_SIMPLE_SEND_RING_BUFFER_TO_SEND
+#endif
+
+
+
+
+#endif //FLOWINSTRUMENTATION
+
```
Comment (igor-ivanov) [inline src/vma/util/flow_instrumentation.h]: @guy101 all of this macros have very long names. I do not see any obvious reason not to use short and simple names for this purpose. Anyone can look at prefix to understand that it relates profiling and read next line of the code. 
Please consider short names. You have description as FLOW_RING_SIMPLE_SEND_RING_BUFFER_TO_SEND and "send_ring_buffer() to send"
For example:
```
#ifdef FLOW_RING_SIMPLE_SEND_RING_BUFFER_TO_SEND
        #define INSTRUMENT_START_131  VMA_TIME_IBPROF_START(131, "send_ring_buffer() to send");
        #define INSTRUMENT_END_131     VMA_TIME_IBPROF_END(131);
#else
        #define INSTRUMENT_START_131
        #define INSTRUMENT_END_131
#endif
```

**Comment by OphirMunk** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/536/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/525/ for details (Mellanox internal link).

**Comment by guy101** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/537/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/526/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/538/ for details (Mellanox internal link).

**Comment by guy101** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/568/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/573/ for details (Mellanox internal link).

File: src/vma/util/flow_instrumentation.h (hunk 2 above)
Comment (guy101) [inline src/vma/util/flow_instrumentation.h]: alex: done
igor: I prefer seperation

File: src/vma/util/flow_instrumentation.h (hunk 3 above)
Comment (guy101) [inline src/vma/util/flow_instrumentation.h]: corrected

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/569/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/574/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/570/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/575/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/571/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/576/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/572/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/577/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/574/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/579/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/575/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/580/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/576/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/581/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/597/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/598/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/599/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/600/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/601/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/602/ for details (Mellanox internal link).

**Comment by liranoz12** [comment]
Currently, There is a compilation error while configure VMA with --enable-time_measure. This #299 PR should fix it.

**Comment by liranoz12** [comment]
Same as #50:

Measuring performance using libibprof cause the total latency to raise. this raise makes this measuringapproch to be prablematic.
Anyway, we have another options in the cause to measure duration of code sections.

**Pull request closed without merging**